crosshair/libimpl/builtinslib.py

```diff
@@ -1937,6 +1937,29 @@ def find(self, substr, start=None, end=None):
         start = 0 if start is None else force_to_smt_sort(start, SmtInt)
         return SmtInt(z3.IndexOf(smt_mystr, smt_substr, start))
 
+    def rfind(self, substr, start=None, end=None):
+        space = self.statespace
+        sub = force_to_smt_sort(substr, SmtStr)
+
+        start = 0 if start is None else force_to_smt_sort(start, SmtInt)
+        end = z3.Length(self.var) if end is None else force_to_smt_sort(end, SmtInt)
+        value = z3.SubString(self.var, start, end - 1)
```

Comment:
Sounds like it. Will include a bounds check. Though I think `find` also has this problem.

Reply:
Ah yeah, three thoughts.
(1) Looks to me like Python essentially behaves as though start and end are clamped to the string bounds; and I think we can do the same here; we just have to be careful that whatever we pass to z3.SubString is clamped appropriately.
(2) It might be a little easier to do this preprocessing in regular python, before getting z3 values out; maybe something like this:
```
mylen = len(self)
if start > mylen or end < 0:
  return -1
if start < 0:
  start = 0
if end > mylen:
  end = mylen
start = force_to_smt_sort(start, SmtInt)
end = force_to_smt_sort(end, SmtInt)
value = z3.SubString(self.var, start, end - start)
```
(3) something I hadn't noticed before; the last argument to z3.SubString is a length, and an ending index, so I think we want `end - start` there.

Reply:
(1) Negative arguments to `find` are supported. `'foo'.find('o', -1)` will return `2`. [According to the documentation](https://docs.python.org/3/library/stdtypes.html#str.find), the semantics are `s.find(sub, start, end) = s[start:end].find(sub) + start`.  
(2) Alright! With this it's also a lot easier to implement the behaviour above.  
(3) and (4) Already have an offset in there ;) I'll also have a look at `find`, see if it needs anything.

Reply:
Ah! I hadn't noticed that it was supposed to act like a regular slice. Good idea! Yes - just slice the string normally.

This is common in CrossHair - it's almost always better to write things in terms of stuff that already works with symbolics.

Reply:
Some odd boundary conditions still, like `z3.SubStr` is ill-defined with negative length argument, but overall code is looking much cleaner now. I also took the liberty of adjusting `find` because it was returning the wrong thing on negative arguments.